pkg/api/endpoints/util.go

```diff
@@ -34,70 +34,93 @@ import (
 // form for things like comparison.  The result is a newly allocated slice.
 func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
```

Comment:
// RepackSubsets takes a slice of EndpointSubset objects, expands it to the full
// representation, and then repacks that into the canonical layout.  This
// ensures that code which operates on these objects can rely on the common
// form for things like comparison.  The result is a newly allocated slice.
func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
	// First map each unique port definition to the sets of hosts that
	// offer it.
	allAddrs := map[addressKey]*api.EndpointAddress{}
	portToAddrReadyMap := map[api.EndpointPort]addressReadyMap{}
	for i := range subsets {
		for j := range subsets[i].Ports {
			epp := subsets[i].Ports[j]
			for k := range subsets[i].Addresses {
				joinAddress(&subsets[i].Addresses[k], epp, true, allAddrs, portToAddrReadyMap)
			}
			for k := range subsets[i].NotReadyAddresses {
				joinAddress(&subsets[i].NotReadyAddresses[k], epp, false, allAddrs, portToAddrReadyMap)
			}
		}
	}

	// Next, map the sets of hosts to the sets of ports they offer.
	// Go does not allow maps or slices as keys to maps, so we have
	// to synthesize an artificial key and do a sort of 2-part
	// associative entity.
	type keyString string
	keyToAddrReadyMap := map[keyString]addressReadyMap{}
	addrReadyMapKeyToPorts := map[keyString][]api.EndpointPort{}
	for epp, addrReadyMap := range portToAddrReadyMap {
		addrReadyMapKey := keyString(hashAddresses(addrReadyMap))
		keyToAddrReadyMap[addrReadyMapKey] = addrReadyMap
		addrReadyMapKeyToPorts[addrReadyMapKey] = append(addrReadyMapKeyToPorts[addrReadyMapKey], epp)
	}

	// Next, build the N-to-M association the API wants.
	final := []api.EndpointSubset{}
	for addrReadyMapKey, ports := range addrReadyMapKeyToPorts {
		var addrs, notReadyAddrs []api.EndpointAddress
		for address, addrReady := range keyToAddrReadyMap[addrReadyMapKey] {
			if addrReady {
				addrs = append(addrs, *address)
			} else {
				notReadyAddrs = append(notReadyAddrs, *address)
			}
		}
		final = append(final, api.EndpointSubset{Addresses: addrs, NotReadyAddresses: notReadyAddrs, Ports: ports})
	}

	// Finally, sort it.
	return SortSubsets(final)
}